sam.c

```diff
@@ -1599,11 +1602,19 @@ static bool cmd_write(Vis *vis, Win *win, Command *cmd, const char *argv[], Sele
 			vis_info_show(vis, "Rejected write to stdout by pre-save hook");
 			return false;
 		}
+		/* a pre-save hook may have changed the text; need to re-take the range */
+		range_all = text_object_entire(text, 0);
```

Comment:
how about:
```
if (write_entire_file)
    *r = text_range_new(0, text_size(text));
```
and leaving the loop body as is?

The same comment also applies to the other hunk.

Reply:
The thing is, it'll of course modify the caller's original range and I'm not sure if this will cause any side effects.

Reply:
Ah wait, did you mean to do it on purpose, as if to somehow signal the caller that the range has changed?

Reply:
Yeah I would do it on purpose. Currently it is probably only used by the grouping logic (which has other issues). The looping constructs are other candidates. But as mentioned earlier, changing the content while operating on it doesn't really fit sam's model.

Reply:
Ok. Done. Also, added a doc comment for `cmd_write`.
I wish `cmd_write` took `Filerange r` instead of `Filerange *r` - not a big fan of mutating args like that.

Anyway, I briefly checked and it seems the range passed to `cmd_write` isn't used anywhere afterwards. Also, a similar thing already happens in `sam_execute` at https://github.com/martanne/vis/blob/d2407f9ebc089e11eb12054f614a37ba6dbcb1ef/sam.c#L1120 so I guess it shouldn't introduce new unwanted side effects. I was just worried about potential regressions in the future.